lib/ansible/plugins/connection/__init__.py

```diff
@@ -269,7 +273,33 @@ def __getattr__(self, name):
                         return method
             raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
 
+    def _connect(self, *args, **kwargs):
+        pass
+
+    def _ensure_socket_path(self):
+        self._update_connection_state()
+
+        if (sp := getattr(self, '_socket_path', '')) and os.path.exists(sp):
+            # connection process should already be running, just talk to it
+            conn = AnsibleConnectionClient(self._socket_path)
+
+            # FIXME: this is trashing the connection details somehow (eg, remote_user is incorrect)- maybe a race where we need to copy it earlier?
+            # is this even necessary anymore now that we're passing the connection options over directly?
+            # pcdata = cPickle.dumps(self._play_context.serialize(), protocol=0)
+            # conn.update_play_context(to_text(pcdata))
+
```

Comment:
The problem is that you start the connection before we know the connection details for sure, so getting the data earlier will make it even worse. In some cases (when all is known before task templating) you should have no issues as play_context carries the correct info, but if you need info to change at the task, or worse, loop/delegation levels you'll run into trouble much sooner.

Most of the time this problem can be ignored as plays tend to reuse the same information w/o changing it, specially networking ones. 

As you note, since we are passing the connection information and using set_options/get_option on them, we should be able to ignore most of the play_context data, but this requires all connection plugins to correctly source the data (all those in core do, but many of those in collections do not). Why I was planning to add deprecation notices to play_context soon.